File: backend/app/routers/favorites.py
"""收藏题目 API 路由"""

import json
from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from ..db import get_db
from ..services.favorite_service import FavoriteService


router = APIRouter(prefix="/api/questions/favorites", tags=["favorites"])


# ===== Request/Response Schemas =====


class AddFavoriteRequest(BaseModel):
    tenant_id: int
    user_id: int
    question_id: int
    question_type_id: int | None = None
    subject_id: int | None = None
    tag_ids: list[int] | None = None


class FavoriteQuestionOut(BaseModel):
    """收藏列表中的题目信息"""

    id: int
    document_id: int
    sequence_index: int
    content: str
    legend_images: List[str] = Field(default_factory=list)
    page: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class QuestionTypeOut(BaseModel):
    """题型信息"""

    id: int
    name: str


class SubjectOut(BaseModel):
    """科目信息"""

    id: int
    name: str


class TagOut(BaseModel):
    """标签信息"""

    id: int
    name: str


class FavoriteItemOut(BaseModel):
    """收藏列表项"""

    id: int
    question_id: int
    question: FavoriteQuestionOut
    question_type: Optional[QuestionTypeOut] = None
    subject: Optional[SubjectOut] = None
    tags: List[TagOut] = Field(default_factory=list)
    created_at: datetime


class AddFavoriteResponse(BaseModel):
    id: int
    question_id: int
    created_at: datetime


class RemoveFavoriteResponse(BaseModel):
    success: bool


class CheckFavoriteResponse(BaseModel):
    is_favorited: bool


class FavoriteQuotaResponse(BaseModel):
    max_favorites: int
    current_count: int
    remaining: int


class FavoritesListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[FavoriteItemOut]


# ===== Helper Functions =====


def _parse_legend_images(legend_json: Optional[str]) -> List[str]:
    """解析 legend_images JSON 字符串"""
    if not legend_json:
        return []
    try:
        val = json.loads(legend_json)
        if isinstance(val, list):
            return [str(x) for x in val]
    except Exception:
        pass
    return []


def _build_favorite_item(favorite) -> FavoriteItemOut:
    """构建收藏列表项响应"""
    question = favorite.question
    
    # 构建题型信息
    question_type = None
    if favorite.question_type:
        question_type = QuestionTypeOut(
            id=favorite.question_type.id,
            name=favorite.question_type.name,
        )
    
    # 构建科目信息
    subject = None
    if favorite.subject:
        subject = SubjectOut(
            id=favorite.subject.id,
            name=favorite.subject.name,
        )
    
    # 构建标签列表
    tags = [
        TagOut(id=tag.id, name=tag.name)
        for tag in favorite.tags
    ]
    
    return FavoriteItemOut(
        id=favorite.id,
        question_id=favorite.question_id,
        question=FavoriteQuestionOut(
            id=question.id,
            document_id=question.document_id,
            sequence_index=question.sequence_index,
            content=question.content,
            legend_images=_parse_legend_images(question.legend_images),
            page=question.page,
            created_at=question.created_at,
            updated_at=question.updated_at,
        ),
        question_type=question_type,
        subject=subject,
        tags=tags,
        created_at=favorite.created_at,
    )


# ===== API Endpoints =====


@router.post("", response_model=AddFavoriteResponse)
def add_favorite(
    payload: AddFavoriteRequest,
    db: Session = Depends(get_db),
):
    """收藏题目
    
    错误码:
    - 404: 题目不存在
    - 409: 已收藏
    - 402: 超过配额
    - 403: 无权访问题型/科目/标签
    """
    service = FavoriteService(db)
    favorite = service.add_favorite(
        tenant_id=payload.tenant_id,
        user_id=payload.user_id,
        question_id=payload.question_id,
        question_type_id=payload.question_type_id,
        subject_id=payload.subject_id,
        tag_ids=payload.tag_ids,
    )
    return AddFavoriteResponse(
        id=favorite.id,
        question_id=favorite.question_id,
        created_at=favorite.created_at,
    )


@router.get("/quota", response_model=FavoriteQuotaResponse)
def get_favorite_quota(
    tenant_id: int = Query(..., description="租户 ID"),
    user_id: int = Query(..., description="用户 ID"),
    db: Session = Depends(get_db),
):
    """获取收藏配额信息"""
    service = FavoriteService(db)
    quota = service.get_quota(tenant_id=tenant_id, user_id=user_id)
    return FavoriteQuotaResponse(
        max_favorites=quota["max_favorites"],
        current_count=quota["current_count"],
        remaining=quota["remaining"],
    )


@router.get("/{question_id}/check", response_model=CheckFavoriteResponse)
def check_favorite(
    question_id: int,
    tenant_id: int = Query(..., description="租户 ID"),
    user_id: int = Query(..., description="用户 ID"),
    db: Session = Depends(get_db),
):
    """检查题目是否已收藏"""
    service = FavoriteService(db)
    is_favorited = service.is_favorited(
        tenant_id=tenant_id,
        user_id=user_id,
        question_id=question_id,
    )
    return CheckFavoriteResponse(is_favorited=is_favorited)


@router.delete("/{question_id}", response_model=RemoveFavoriteResponse)
def remove_favorite(
    question_id: int,
    tenant_id: int = Query(..., description="租户 ID"),
    user_id: int = Query(..., description="用户 ID"),
    db: Session = Depends(get_db),
):
    """取消收藏
    
    错误码:
    - 404: 收藏记录不存在
    """
    service = FavoriteService(db)
    service.remove_favorite(
        tenant_id=tenant_id,
        user_id=user_id,
        question_id=question_id,
    )
    return RemoveFavoriteResponse(success=True)


@router.get("", response_model=FavoritesListResponse)
def get_favorites(
    tenant_id: int = Query(..., description="租户 ID"),
    user_id: int = Query(..., description="用户 ID"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    db: Session = Depends(get_db),
):
    """获取收藏列表（分页）"""
    service = FavoriteService(db)
    total, favorites = service.get_favorites(
        tenant_id=tenant_id,
        user_id=user_id,
        page=page,
        page_size=page_size,
    )
    return FavoritesListResponse(
        total=total,
        page=page,
        page_size=page_size,
        items=[_build_favorite_item(f) for f in favorites],
    )

File: backend/src/routes/favorites.ts
import { Hono } from "hono"
import { z } from "zod"
import { FavoritesService } from "../domains/favorites/service"
import { requireAuth } from "./auth-context"

const addFavoriteSchema = z.object({
  questionID: z.string().min(1),
  questionTypeID: z.string().min(1).nullable().optional(),
  subjectID: z.string().min(1).nullable().optional(),
  tagIDs: z.array(z.string().min(1)).optional(),
})

export const favoritesRoutes = new Hono()

favoritesRoutes.get("/", async (c) => {
  const { user } = await requireAuth(c)
  const page = c.req.query("page") ? Number(c.req.query("page")) : 1
  const pageSize = c.req.query("page_size") ? Number(c.req.query("page_size")) : 20
  if (!Number.isInteger(page) || page < 1) throw new Error("Invalid page")
  if (!Number.isInteger(pageSize) || pageSize < 1 || pageSize > 100) throw new Error("Invalid page_size")

  const result = await FavoritesService.list({
    userID: user.id,
    page,
    pageSize,
  })

  return c.json({
    total: result.total,
    page,
    page_size: pageSize,
    items: result.items.map((item) => ({
      id: item.favorite.id,
      question_id: item.favorite.questionID,
      question: {
        id: item.question.id,
        document_id: item.question.sourceDocumentID ?? item.question.studioDocumentID,
        studio_document_id: item.question.studioDocumentID,
        source_document_id: item.question.sourceDocumentID ?? null,
        sequence_index: item.question.sequenceIndex,
        knowledge_title: item.knowledgeTitle,
        content: item.question.content,
        legend_images: item.question.legendImages,
        page: item.question.page,
        created_at: item.question.createdAt,
        updated_at: item.question.updatedAt,
      },
      studio_question_card_id: item.studioQuestionCardID,
      question_type: item.questionType
        ? {
            id: item.questionType.id,
            name: item.questionType.name,
            created_at: item.questionType.createdAt,
          }
        : null,
      subject: item.subject
        ? {
            id: item.subject.id,
            name: item.subject.name,
            created_at: item.subject.createdAt,
          }
        : null,
      tags: item.tags.map((tag) => ({
        id: tag.id,
        name: tag.name,
        created_at: tag.createdAt,
      })),
      created_at: item.favorite.createdAt,
    })),
  })
})

favoritesRoutes.get("/quota", async (c) => {
  const { user } = await requireAuth(c)
  const quota = await FavoritesService.quota({ userID: user.id })
  return c.json({
    used: quota.used,
    limit: quota.limit,
    remaining: quota.remaining,
  })
})

favoritesRoutes.post("/", async (c) => {
  const { user } = await requireAuth(c)
  const body = addFavoriteSchema.parse(await c.req.json())
  const favorite = await FavoritesService.add({
    userID: user.id,
    questionID: body.questionID,
    questionTypeID: body.questionTypeID,
    subjectID: body.subjectID,
    tagIDs: body.tagIDs,
  })
  return c.json(
    {
      id: favorite.id,
      question_id: favorite.questionID,
      question_type_id: favorite.questionTypeID ?? null,
      subject_id: favorite.subjectID ?? null,
      tag_ids: favorite.tagIDs,
      created_at: favorite.createdAt,
    },
    201,
  )
})

favoritesRoutes.get("/:questionID/check", async (c) => {
  const { user } = await requireAuth(c)
  return c.json({
    is_favorited: await FavoritesService.isFavorited({
      userID: user.id,
      questionID: c.req.param("questionID"),
    }),
  })
})

favoritesRoutes.delete("/:questionID", async (c) => {
  const { user } = await requireAuth(c)
  await FavoritesService.remove({
    userID: user.id,
    questionID: c.req.param("questionID"),
  })
  return c.json({ success: true })
})
